pkg/syncer/obsctlsyncer.go

```diff
@@ -0,0 +1,354 @@
+	audience       string
+	issuerURL      string
+	managedTenants string
+
+	autoDetectSecretsFn func(ctx context.Context,
+		k8s client.Client,
+		namespace, audience, issuerURL, managedTenants string,
+	) (map[string]*config.OIDCConfig, error)
+
+	c *config.Config
+
+	lokiRulesSetOps      *prometheus.CounterVec
+	promRulesSetOps      *prometheus.CounterVec
+	lokiRulesSetFailures *prometheus.CounterVec
+	promRulesSetFailures *prometheus.CounterVec
+}
+
+func NewObsctlRulesSyncer(
+	ctx context.Context,
+	logger log.Logger,
+	kc client.Client,
+	namespace, apiURL, audience, issuerURL, managedTenants string,
+	reg prometheus.Registerer,
+) *ObsctlRulesSyncer {
+	return &ObsctlRulesSyncer{
+		ctx:            ctx,
+		logger:         logger,
+		k8s:            kc,
+		apiURL:         apiURL,
+		namespace:      namespace,
+		audience:       audience,
+		issuerURL:      issuerURL,
+		managedTenants: managedTenants,
+
+		autoDetectSecretsFn: AutoDetectTenantSecrets,
+
+		lokiRulesSetOps: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_loki_rule_sets_total",
+			Help: "Total number of obsctl set operations for lokiv1/v1beta1 rules.",
+		}, []string{"type", "tenant"}),
+		promRulesSetOps: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_prom_rule_sets_total",
+			Help: "Total number of obsctl set operations for monitoringv1 rules.",
+		}, []string{"tenant"}),
+		lokiRulesSetFailures: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_loki_rule_set_failures_total",
+			Help: "Total number of failed obsctl set operations for lokiv1/v1beta1 rules.",
+		}, []string{"type", "tenant"}),
+		promRulesSetFailures: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_prom_rule_set_failures_total",
+			Help: "Total number of failed obsctl set operations for monitoringv1 rules.",
+		}, []string{"tenant"}),
+	}
+}
+
+func AutoDetectTenantSecrets(
+	ctx context.Context,
+	k8s client.Client,
+	namespace, audience, issuerURL, managedTenants string,
+) (map[string]*config.OIDCConfig, error) {
+	tenantSecret := map[string]*config.OIDCConfig{}
+
+	// List secrets.
+	secret := corev1.SecretList{}
+	if err := k8s.List(ctx, &secret, client.InNamespace(namespace)); err != nil {
+		return nil, err
+	}
+
+	// Filter secrets for configured tenants.
+	configuredTenants := strings.Split(managedTenants, ",")
+	for i := range secret.Items {
+		lbls := secret.Items[i].Labels
+
+		if _, ok := lbls["tenant"]; !ok {
+			continue
+		}
+
+		// If tenant is not configured, skip.
+		if !slices.Contains(configuredTenants, lbls["tenant"]) {
+			continue
+		}
+
+		if secret.Items[i].Data == nil {
+			continue
+		}
+
+		tOIDC := &config.OIDCConfig{
+			Audience:      audience,
+			IssuerURL:     issuerURL,
+			OfflineAccess: false,
+		}
+
+		// Get tenant credentials from secret.
+		// TODO: Define spec for secrets. Currently can be both underscore and dash.
```

Comment:
Agree here, lets just document this format here and be strict about it. Either one or the other.

Reply:
By here, do you mean on this repo's readme? Or in tenant docs? 🙂

Reply:
We can be strict here, but then rolling it out would be blocked as currently, both dash and underscore are being used in RHOBS

Reply:
I think we can document here? (can be a follow-up) Including the label for discovery of secrets and we can cross link with the tenant docs wherever they exist

happy to keep support for both in theory but lets just document a single approach

Reply:
Ack, will do